### DISCUSSION TITLE
Plunge watering can created with Replicad

### DISCUSSION THREAD

**Discussion by raydeleu** [sgenoud/replicad#35] [Show and tell]
<img width="539" alt="plunge-v6" src="https://user-images.githubusercontent.com/38007983/174858404-ee5f39ba-4d5c-4db0-8d36-e536db95e295.png">

I created this "Plunge" water carafe for plants designed by Robert Bronwasser (see https://www.robertbronwasser.com/project/spring/) in Replicad. It is not an exact copy but demonstrates the capabilities and partly also the limitations of the library. The model consists of three main parts which I called the "body", the "filler" and the "spout". The cone-like body is created as a body of revolution. The filler is built as a loft between three wires, where the middle wire coincides with the top of the body. The spout is a small cylinder at an angle. The shapes are combined in a boolean fuse operation and then filleted. I tried to create a hollow shape by identifying the filler opening and the spout opening, but I could not get that to work. So in the end I created a shell by removing only the opening of the spout. The filler was then opened using a "cutter". 

The code and some remarks regarding kernel errors are shown below"

```` javascript
 // Model of the Plunge watering carafe designed by Robert Bronwasser

function main({
                Sketcher,
                sketchCircle,
                sketchFaceOffset,
                makeCylinder,
                sketchRectangle,
                FaceFinder,
            })
{

// side profile of the bottom of the carafe
let p0 = [0,0]
let p1 = [20,0]
let p2 = [30,5]
let p3 = [30,8]
let p4 = [8,100]   // radius of the top at 100 mm is 8 mm
let p5 = [0,100]

let sideview = new Sketcher("XZ")
.lineTo(p1)
.lineTo(p2)
.lineTo(p3)
.lineTo(p4)
.lineTo(p5)
.close()

// sketch is created on XZ plane, revolve is per default around z-axis
let body = sideview.revolve()

// create cross sections of the filler for the carafe
//          used a workaround to rotate and translate the sketch to the required position
let fillHole = sketchCircle(12).face().rotate(-20,[0,0,0],[0,1,0]).translate([-35,0,135])
fillHole = sketchFaceOffset(fillHole,0);
let topBody = sketchCircle(8).face().translate([0,0,100]);   // radius 8 at 100 mm 
topBody = sketchFaceOffset(topBody,0); 
let fillBottom = sketchCircle(9).face().rotate(20,[0,0,0],[0,1,0]).translate([0,0,80]); 
fillBottom = sketchFaceOffset(fillBottom,0); 

// filler shape is created as a loft between the three wires
let filler    = fillHole.loftWith([topBody,fillBottom],{ruled: false});

// create spout, a cylinder with radius 5, length "lengthSpout"
let angleSpout = 45
let lengthSpout = 70
let spout = makeCylinder(5,lengthSpout,[0,0,0],[0,0,1]).rotate(angleSpout,[0,0,0],[0,1,0]).translate([0,0,100])

// // union the main body with the filler and fillet the junction with radius 30
let plunge = body.fuse(filler);
plunge = plunge.fillet(30,(e)=>e.inPlane("XY",100)) 

// union the shape with the sprout and fillet the junction with radius 10
plunge = plunge.fuse(spout).fillet(10,(e)=>e.inBox([20,20,100],[-20,-20,120]));

// Create a shell 
let pointFiller = [-35,0,135]
let spoutOpening = []
spoutOpening[0] = Math.cos(angleSpout*Math.PI/180)*lengthSpout
spoutOpening[1] = 0
spoutOpening[2] = 100 + Math.sin(angleSpout*Math.PI/180)*lengthSpout

const orifices = new FaceFinder().either([
    (f) => f.containsPoint(pointFiller),
    (f) => f.containsPoint(spoutOpening)]);

// plunge = plunge.shell(-0.5,(f)=>f.inBox([20,20,130],[[-20,-20,155]])); // Kernel Error
// plunge = plunge.shell(-0.5,(f)=>f.containsPoint(pointFiller)); // works!
// plunge = plunge.shell(-0.5,orifices); // Kernel Error
plunge = plunge.shell(-1,(f)=>f.containsPoint(spoutOpening)); // works!

// create cutter boxes 
//      approach to open up another face by subtracting a box from the shape
let cutterFiller = sketchRectangle(40,30).extrude(20).rotate(-20,[0,0,0],[0,1,0]).translate([-30,0,135])
// let cutterSpout = sketchRectangle(40,30).extrude(20).rotate(10,[0,0,0],[0,1,0]).translate([48,0,145])

plunge = plunge.cut(cutterFiller)
// plunge = plunge.cut(cutterSpout)  // resulted in Kernel error without the shell command

// let plungeInner = makeOffset(plunge,-1)  // resulted in Kernel error
// let hollowPlunge = plunge.cut(plungeInner)  // not possible as makeOffset failed

let shapeArray = [
{shape: plunge, name: "plunge", color:"steelblue"}, 
// {shape: fillHole, color:"black"},   // note that after the loft these sketches are deleted? 
// {shape: topBody, color:"black"},
// {shape: fillBottom, color: "black"},
// {shape: filler, color: "yellow"},
// {shape: sprout, color: "blue"},
// {shape: cutterFiller},
// {shape:cutterSpout},
// {shape: test}
]

return shapeArray
}

````

**Comment by raydeleu** [comment]
<img width="337" alt="plunge_parts_2,png" src="https://user-images.githubusercontent.com/38007983/175698315-75023470-fa37-4d2c-853d-b06dadee22d1.png">

Using the new option to display a part with opacity I created this image that illustrates some parts that were used to construct the final shape. This option can be very useful to understand how parts are overlapping and why a boolean operation between the shapes could fail. The code above was adapted like this: 

```` javascript 
let shapeArray = [
//{shape: plunge, name: "plunge", color:"steelblue", opacity: 0.5}, 
// {shape: fillHole, color:"black"},   // note that after the loft these sketches are deleted? 
// {shape: topBody, color:"black"},
// {shape: fillBottom, color: "black"},
{shape: body, color: "orange",opacity:0.5},
{shape: filler, color: "red"},
{shape: spout, color: "blue"},
{shape: cutterFiller, color: "green", opacity: 0.5},
// {shape:cutterSpout},
// {shape: test}
]
````

**Comment by sgenoud** [comment]
I have played a little bit with your model, and modified it to how I think it is the easiest to do now, have a look [here](https://studio.replicad.xyz/workbench?code=UEsDBAoAAAAIAGmCSlZwmaNjjAMAAKYIAAAHAAAAY29kZS5qc41V227jNhB991dM9UR1ZUWy16ibIH3IAkULtNigLdBsjDww0ihWlyYFkrKTLvzvHV5ky062WCCIzNHwzJkzF1VKGgtfJgC15rssPv%252FAynL5JHAwfGh1FU4b%252FhlvBZc42cM1aOxEW%252FH6ajKpPFKNDe%252BFveWabww5fDFrtaOTNZfQcGFwP5k0vaxsqyRhtZLpbOS0T4kLRbm4ANPWCJ1WTSsQVAN2jfCorFWb4VRRkAbJW6D17tsWdxTTEWYp2QHy9W%252BtRDYr4lG4U1lksIiGrX8%252FH73%252BS7HVMoOyKB5OMKbLeKyEMkj4keZntNUaWgOVRm6xBkrs7h46p1FGAm2V2KJ736Ee5AGuVS9r%252BHfKn1sTM3hU9QuxHxLJA%252FJH6dVmyd19kuYR7hA9xKSHMgYMelnNoA8pJyhmo%252FSpXKFS7u0vSoRiUtxDYVmad%252B1WWTadkVDJJwprNZdGUCS2ms4XGZC9nC%252B8Po75ABW1D83Cyll6lsNJSJ%252BCu21VdxNSH11ent9N7j4lvirpFfjUNa%252Fb3sASuHVm2GzgJLcb3ytfz%252B51cvdsWYxTCginvH58K6dRqKEwUXuz5h2Om4Mb4CBUQ%252BVGu0OUvjR2rRFjK6EZUSCI64O8ubv3d2vXbBU1y0Y8H2iOQPcC6zhpsE%252BvTtvEdKq3GRGoXqjTawLfEdqgJNVVoHwiQxKefzr3JLLx68BbiNH7RbSOHMn8Q3GogA%252F1dfGPaK87rAj9FefPz3bEH9VhcV6HY8A0x2er%252BxrZiN1QF%252Frrpds9Tna3f8LYeR1GM8NpOP1P663%252FDBtrrNe8iPQ60csnl6eDypvehK5ATfrD8W34kQdYNqccGaZw%252FRNg3r5q8oHvkWxopQNRQ6uxt99EtCzeouFoetHSgVJ5SulGPbOVm5JZLEcGK78S%252FL9yRvU5NNiH0GCcWKIQMAijWml%252FHtr4ZHWMK%252FuxQ9nKJ%252Bfi7GPbqngg8%252B%252FcrvNKmVHbwPfBevsrXEBJc0uGUbnPYUoHU5xbZ87qdse7AGboc%252FTtId4Q1WfPpiXNpVOxyZHKgJqt%252FJeDNWmw0ohYaj1z6%252FRhI5XS7P8cx9zTh9AibQPsu8PnM%252FVfcqCvju21pH3gm%252BYy0stA8g2dknCkTquUUJoMxiKKR9FjAnsC2DvkiBGoH5Ae%252FdYZ7imaWg%252BkOl619uUSinwB%252B%252Bz0ThiG4y2NdfLKKe6mwSeScU4Pk%252F1%252FUEsBAhQACgAAAAgAaYJKVnCZo2OMAwAApggAAAcAAAAAAAAAAAAAAAAAAAAAAGNvZGUuanNQSwUGAAAAAAEAAQA1AAAAsQMAAAAA).

Next step is to modify the official doc to teach people the tools I use here!
